crates/matrix-sdk/src/widget/client/matrix/mod.rs

```diff
@@ -0,0 +1,217 @@
+    room::{MessagesOptions, Room as Joined},
+    widget::{
+        messages::{
+            from_widget::{
+                ReadEventRequest, ReadEventResponse, SendEventRequest, SendEventResponse,
+            },
+            EventType, MatrixEvent, OpenIDRequest, OpenIDState,
+        },
+        EventFilter, Permissions, PermissionsProvider,
+    },
+};
+
+#[derive(Debug)]
+pub struct Driver<T> {
+    room: Joined,
+    permissions_provider: T,
+    event_handler_handle: Option<EventHandlerDropGuard>,
+}
+
+impl<T> Driver<T> {
+    pub fn new(room: Joined, permissions_provider: T) -> Self {
+        Self { room, permissions_provider, event_handler_handle: None }
+    }
+}
+
+#[async_trait]
+impl<T: PermissionsProvider> Client for Driver<T> {
+    async fn initialise(&mut self, permissions: Permissions) -> Capabilities {
+        let permissions = self.permissions_provider.acquire_permissions(permissions).await;
+
+        Capabilities {
+            listener: Filters::new(permissions.read.clone())
+                .map(|filters| self.setup_event_listener(filters)),
+            reader: Filters::new(permissions.read).map(|filters| {
+                let reader: Box<dyn Reader> =
+                    Box::new(EventServerProxy::new(self.room.clone(), filters));
+                reader
+            }),
+            sender: Filters::new(permissions.send).map(|filters| {
+                let sender: Box<dyn Sender> =
+                    Box::new(EventServerProxy::new(self.room.clone(), filters));
+                sender
+            }),
+        }
+    }
+
+    fn get_openid(&self, req: OpenIDRequest) -> OpenIDStatus {
+        let user_id = self.room.own_user_id().to_owned();
+        let client = self.room.client.clone();
+        let (tx, rx) = oneshot::channel();
+        tokio::spawn(async move {
+            let _ = tx.send(
+                client
+                    .send(MatrixOpenIDRequest::new(user_id), None)
+                    .await
+                    .map(|res| {
+                        OpenIDDecision::Allowed(OpenIDState {
+                            id: req.id,
+                            token: res.access_token,
+                            expires_in_seconds: res.expires_in.as_secs() as usize,
+                            server: res.matrix_server_name.to_string(),
+                            kind: res.token_type.to_string(),
+                        })
+                    })
+                    .unwrap_or(OpenIDDecision::Blocked),
+            );
+        });
+
+        // TODO: getting an OpenID token generally has multiple phases as per the JS
+        // implementation of the `ClientWidgetAPI`, e.g. the `MatrixDriver`
+        // would normally have some state, so that if a token is requested
+        // multiple times, it may return/resolve the token right away.
+        // Currently, we assume that we always request a new token. Fix it later.
+        OpenIDStatus::Pending(rx)
```

Comment:
we should also mention in this todo that we do not have any hooks for the app to prompt the user for consent. This is not required since we just auto approve for EC

Reply:
Should it be part of this particular implementation or rather a different place? (something which would prompt the user) Also, do we really want to mention Element Call explicitly there (from my understanding, it was meant to be a generic implementation without some Element Call specifics?).